src/app/zap-templates/partials/im_command_handler_cluster_commands.zapt

```diff
@@ -8,40 +9,77 @@ const uint8_t * {{asSymbol label}};
+      {
+            ChipLogProgress(Zcl, "Ignoring duplicate TLV tag %" PRIx32, TLV::TagNumFromTag(dataTlv.GetTag()));
+            break;
+      }
 {{#if (isOctetString type)}}
     {
         const uint8_t * data = nullptr;
-        TLVError = dataTlv.GetDataPtr(data);
+        TLVUnpackError = dataTlv.GetDataPtr(data);
         {{asSymbol label}} = chip::ByteSpan(data, dataTlv.GetLength());
     }
 {{else if (isString type)}}
-      TLVError = dataTlv.GetDataPtr({{asSymbol label}});
+      // TODO(#5542): The cluster handlers should accept a ByteSpan for all string types.
+      TLVUnpackError = dataTlv.GetDataPtr({{asSymbol label}});
 {{else if isArray}}
       // Just for compatibility, we will add array type support in IM later.
-      TLVError = dataTlv.GetDataPtr(const_cast<const uint8_t *&>({{asSymbol label}}));
+      TLVUnpackError = dataTlv.GetDataPtr(const_cast<const uint8_t *&>({{asSymbol label}}));
 {{else}}
-      TLVError = dataTlv.Get({{asSymbol label}});
+      TLVUnpackError = dataTlv.Get({{asSymbol label}});
 {{/if}}
+      if (CHIP_NO_ERROR == TLVUnpackError)
+      {
+            {{asSymbol label}}Exists = true;
+            validArgumentCount++;
+      }
       break;
 {{/zcl_command_arguments}}
     default:
       // Unsupported tag, ignore it.
       ChipLogProgress(Zcl, "Unknown TLV tag during processing.");
       break;
   }
-  if (TLVError != CHIP_NO_ERROR)
+  if (TLVUnpackError != CHIP_NO_ERROR)
   {
-    ChipLogProgress(Zcl, "Failed to decode TLV data with tag %" PRIx32 ": %" PRId32, TLV::TagNumFromTag(dataTlv.GetTag()), TLVError);
+    ChipLogProgress(Zcl, "Failed to decode TLV data with tag %" PRIx32 ": %" PRId32, TLV::TagNumFromTag(dataTlv.GetTag()), TLVUnpackError);
+    break;
   }
 }
+
+if (CHIP_END_OF_TLV == TLVError)
+{
+      // CHIP_END_OF_TLV means we have iterated all items in the structure, which is not a real error.
+      TLVError = CHIP_NO_ERROR;
+}
+else
+{
+      ChipLogProgress(Zcl, "Failed to decode TLV data: %" PRId32, TLVError);
+}
+{{/if}}
+
+{{#if (zcl_command_arguments_count this.id)}}
+if (CHIP_NO_ERROR == TLVError && CHIP_NO_ERROR == TLVUnpackError && {{zcl_command_arguments_count this.id}} == validArgumentCount)
+{
 {{/if}}
 // TODO(#5098) We should pass the Command Object and EndpointId to the cluster callbacks.
 emberAf{{asCamelCased parent.name false}}Cluster{{asCamelCased name false}}Callback({{#zcl_command_arguments}}{{#if (isCharString type)}}const_cast<uint8_t*>({{asSymbol label}}){{else}}{{asSymbol label}}{{/if}}{{#unless (isLastElement index count)}}, {{/unless}}{{/zcl_command_arguments}});
+{{#if (zcl_command_arguments_count this.id)}}
+}
+else if ({{zcl_command_arguments_count this.id}} != validArgumentCount)
+{
+      // Additional log for missing command arguments.
+      ChipLogProgress(Zcl, "Missing command arguments in TLV data, command requires %d, given %" PRIu32, {{zcl_command_arguments_count this.id}}, validArgumentCount);
+}
+{{/if}}
```

Comment:
Somewhere here, don't we need to send an error response if we hit any of those error conditions?  Followup/TODO is probably fine, but please make sure this is tracked.

Reply:
Added a TODO (#5590 ) for this and some other places related to the same issue.